clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -446,7 +441,11 @@ private void handleFetchResponse(ClientResponse resp, FetchRequest request) {
                 TopicPartition tp = entry.getKey();
                 FetchResponse.PartitionData partition = entry.getValue();
                 if (!subscriptions.assignedPartitions().contains(tp)) {
+                    // this can happen when a rebalance happened while fetch is still in-flight
                     log.debug("Ignoring fetched data for partition {} which is no longer assigned.", tp);
+                } else if (!subscriptions.isFetchable(tp)) {
```

Comment:
SubscriptionState.isFetchable also checks that the partition is assigned, so maybe we can remove the assignment check just above?